File: 823-binary-trees-with-factors/823-binary-trees-with-factors.cpp

class Solution {
public:
    int numFactoredBinaryTrees(vector<int>&nums) 
    {
        int ans=0;
        int mod=1e9+7;
        int n=nums.size();
        unordered_map<int,long long>mp;
        sort(nums.begin(),nums.end());
        for(int i=0;i<n;i++)
        {
            mp[nums[i]]=1;
            for(auto &[a,b]:mp)
            {
                if(nums[i]%a==0&&mp.find(nums[i]/a)!=mp.end())
                {
                    long long temp=(1ll*mp[a]*mp[nums[i]/a]);
                    mp[nums[i]]=(mp[nums[i]]+temp)%mod;
                }
            }
            ans=(ans+mp[nums[i]])%mod;
        }
        return ans;
    }
};

File: 823-binary-trees-with-factors/823-binary-trees-with-factors.py
class Solution:
    def numFactoredBinaryTrees(self, nums: List[int]) -> int:
        dp={}
        nums.sort()
        # print(nums)
        for i in range(0,len(nums)):
            dp[nums[i]]=1;
            for j in range(i):
                if nums[i]%nums[j]==0:
                    dp[nums[i]]+=(dp.get(nums[j],0)*dp.get(nums[i]/nums[j],0))
            # print(dp[nums[i]])
                    
        ans=0;
        for a,b in dp.items():
            ans=(ans+b)%(10**9+7)
        return ans
            
        